### ISSUE TITLE
Deutsch-Josza Benchmarking Test is giving throwing Error

### ISSUE THREAD

**Issue by manu123416** [SRI-International/QC-App-Oriented-Benchmarks#127]
Hi Tried, I tried running the below code from your Deush Josza algorithm:
   
```
"""
Deutsch-Jozsa Benchmark Program - Qiskit
"""

import sys
import time

import numpy as np
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister

sys.path[1:1] = ["_common", "_common/qiskit"]
sys.path[1:1] = ["../../_common", "../../_common/qiskit"]
import execute as ex
import metrics as metrics

np.random.seed(0)

verbose = False

# saved circuits for display
QC_ = None
C_ORACLE_ = None
B_ORACLE_ = None


############### Circuit Definition

# Create a constant oracle, appending gates to given circuit
def constant_oracle(input_size, num_qubits):
    # Initialize first n qubits and single ancilla qubit
    qc = QuantumCircuit(num_qubits, name=f"Uf")

    output = np.random.randint(2)
    if output == 1:
        qc.x(input_size)

    global C_ORACLE_
    if C_ORACLE_ == None or num_qubits <= 6:
        if num_qubits < 9: C_ORACLE_ = qc

    return qc


# Create a balanced oracle.
# Perform CNOTs with each input qubit as a control and the output bit as the target.
# Vary the input states that give 0 or 1 by wrapping some of the controls in X-gates.
def balanced_oracle(input_size, num_qubits):
    # Initialize first n qubits and single ancilla qubit
    qc = QuantumCircuit(num_qubits, name=f"Uf")

    b_str = "10101010101010101010"  # permit input_string up to 20 chars
    for qubit in range(input_size):
        if b_str[qubit] == '1':
            qc.x(qubit)

    qc.barrier()

    for qubit in range(input_size):
        qc.cx(qubit, input_size)

    qc.barrier()

    for qubit in range(input_size):
        if b_str[qubit] == '1':
            qc.x(qubit)

    global B_ORACLE_
    if B_ORACLE_ == None or num_qubits <= 6:
        if num_qubits < 9: B_ORACLE_ = qc

    return qc


# Create benchmark circuit
def DeutschJozsa(num_qubits, type):
    # Size of input is one less than available qubits
    input_size = num_qubits - 1

    # allocate qubits
    qr = QuantumRegister(num_qubits);
    cr = ClassicalRegister(input_size);
    qc = QuantumCircuit(qr, cr, name="main")

    for qubit in range(input_size):
        qc.h(qubit)
    qc.x(input_size)
    qc.h(input_size)

    qc.barrier()

    # Add a constant or balanced oracle function
    if type == 0:
        Uf = constant_oracle(input_size, num_qubits)
    else:
        Uf = balanced_oracle(input_size, num_qubits)
    qc.append(Uf, qr)

    qc.barrier()

    for qubit in range(num_qubits):
        qc.h(qubit)

    # uncompute ancilla qubit, not necessary for algorithm
    qc.x(input_size)

    qc.barrier()

    for i in range(input_size):
        qc.measure(i, i)

    # save smaller circuit and oracle subcircuit example for display
    global QC_
    if QC_ == None or num_qubits <= 6:
        if num_qubits < 9: QC_ = qc

    # return a handle to the circuit
    return qc


############### Result Data Analysis

# Analyze and print measured results
# Expected result is always the type, so fidelity calc is simple
def analyze_and_print_result(qc, result, num_qubits, type, num_shots):
    # Size of input is one less than available qubits
    input_size = num_qubits - 1

    # obtain counts from the result object
    counts = result.get_counts(qc)
    if verbose: print(f"For type {type} measured: {counts}")

    # create the key that is expected to have all the measurements (for this circuit)
    if type == 0:
        key = '0' * input_size
    else:
        key = '1' * input_size

    # correct distribution is measuring the key 100% of the time
    correct_dist = {key: 1.0}

    # use our polarization fidelity rescaling
    fidelity = metrics.polarization_fidelity(counts, correct_dist)

    return counts, fidelity


################ Benchmark Loop

# Execute program with default parameters
def run(min_qubits=3, max_qubits=8, max_circuits=3, num_shots=100,
        backend_id='qasm_simulator', provider_backend=None,
        hub="ibm-q", group="open", project="main", exec_options=None):
    print("Deutsch-Jozsa Benchmark Program - Qiskit")

    # validate parameters (smallest circuit is 3 qubits)
    max_qubits = max(3, max_qubits)
    min_qubits = min(max(3, min_qubits), max_qubits)
    # print(f"min, max qubits = {min_qubits} {max_qubits}")

    # Initialize metrics module
    metrics.init_metrics()

    # Define custom result handler
    def execution_handler(qc, result, num_qubits, type, num_shots):

        # determine fidelity of result set
        num_qubits = int(num_qubits)
        counts, fidelity = analyze_and_print_result(qc, result, num_qubits, int(type), num_shots)
        metrics.store_metric(num_qubits, type, 'fidelity', fidelity)

    # Initialize execution module using the execution result handler above and specified backend_id
    ex.init_execution(execution_handler)
    ex.set_execution_target(backend_id, provider_backend=provider_backend,
                            hub=hub, group=group, project=project, exec_options=exec_options)

    # Execute Benchmark Program N times for multiple circuit sizes
    # Accumulate metrics asynchronously as circuits complete
    for num_qubits in range(min_qubits, max_qubits + 1):

        input_size = num_qubits - 1

        # determine number of circuits to execute for this group
        num_circuits = min(2, max_circuits)

        print(f"************\nExecuting [{num_circuits}] circuits with num_qubits = {num_qubits}")

        # loop over only 2 circuits
        for type in range(num_circuits):
            # create the circuit for given qubit size and secret string, store time metric
            ts = time.time()
            qc = DeutschJozsa(num_qubits, type)
            metrics.store_metric(num_qubits, type, 'create_time', time.time() - ts)

            # collapse the sub-circuit levels used in this benchmark (for qiskit)
            qc2 = qc.decompose()

            # submit circuit for execution on target (simulator, cloud simulator, or hardware)
            ex.submit_circuit(qc2, num_qubits, type, num_shots)

        # Wait for some active circuits to complete; report metrics when groups complete
        ex.throttle_execution(metrics.finalize_group)

    # Wait for all active circuits to complete; report metrics when groups complete
    ex.finalize_execution(metrics.finalize_group)

    # print a sample circuit
    print("Sample Circuit:");
    print(QC_ if QC_ != None else "  ... too large!")
    print("\nConstant Oracle 'Uf' =");
    print(C_ORACLE_ if C_ORACLE_ != None else " ... too large or not used!")
    print("\nBalanced Oracle 'Uf' =");
    print(B_ORACLE_ if B_ORACLE_ != None else " ... too large or not used!")

    # Plot metrics for all circuit sizes
    metrics.plot_metrics("Benchmark Results - Deutsch-Jozsa - Qiskit")


# if main, execute method
if __name__ == '__main__': run()
```

**I am getting the below error:**
```
C:\Users\manuc\Documents\Pytorch_Study_Workspace\Benchmark_dj\venv\Scripts\python.exe C:/Users/manuc/Documents/Pytorch_Study_Workspace/Benchmark_dj/main.py
Traceback (most recent call last):
  File "C:\Users\manuc\Documents\Pytorch_Study_Workspace\Benchmark_dj\main.py", line 219, in <module>
    if __name__ == '__main__': run()
  File "C:\Users\manuc\Documents\Pytorch_Study_Workspace\Benchmark_dj\main.py", line 161, in run
    metrics.init_metrics()
AttributeError: module 'metrics' has no attribute 'init_metrics'
Deutsch-Jozsa Benchmark Program - Qiskit

Process finished with exit code 1
```

**Please guide me how to correct this Error?**

**Comment by rtvuser1** [comment]
It looks like you are running this from a local directory ... Pytorch_Study_Workspace\Benchmark_dj

This .py file will not run in isolation.  It requires the __common directory to reside in parallel as it imports files from _common to manage job execution and collect/plot metrics. It was done this way to avoid having to create and install a "benchmark support package".

The best way to run these benchmarks is in the top-level directory of the QC-App-Oriented-Benchmarks repository, using the Jupyter notebook benchmarks-qiskit.ipynb, as described in the README.  This notebook executes all the benchmarks if you do a "Run All". 

Alternatively, you can execute each benchmark in its respective subdirectory, e.g qiskit/deutsch-jozsa with a simple python dj_benchmark.py.

Please let us know if you get past this Issue so that it can be resolved.

**Comment by manu123416** [comment]
I added the commons directory in the project. Getting the below result.
![image](https://user-images.githubusercontent.com/95257431/177908333-a4b9261d-8b48-4c2c-b4b5-eb09c2043b71.png)
The directory structure of my project looks like as below:
![image](https://user-images.githubusercontent.com/95257431/177908741-6de570b4-234a-4b0d-81e1-57d8df6e3f50.png)
I hope this is the expected result.
I only have the access to the Qasm simulator.

**Comment by rtvuser1** [comment]
Yes that looks correct.
Using the simulator, the circuit creation and execution time is very small, the fidelity of execution is high (~1.0) and the circuit depth is relatively constant (for DJ).
However, there should be another plot that gets produced, the volumetric plot showing the circuit profile.

Try the other benchmarks, you will see more interesting results (DJ is very limited in what it does).  Best way to run is in the Jupyter notebook that is at the top level, where everything is configured to run all the benchmarks. (note that it has the extension .template which you need to remove when you copy the file).

You can also reference the READM at top level and this paper for more info:  https://arxiv.org/abs/2110.03137

I will leave this Issue open for a little longer.  Please feel free to ask additional questions here.
I'll close this Issue once you let me know you have a good handle on things.

**Comment by rtvuser1** [comment]
Closing this issue for now

**Comment by rtvuser1** [comment]
Closing for now, as new issue created for new issues
